pkg/cmd/agent-task/shared/log.go

```diff
@@ -0,0 +1,570 @@
+		if err != nil {
+			return fmt.Errorf("failed to marshal JSON: %w", err)
+		}
+
+		if label != "" {
+			if err := renderRawMarkdown(label, w, io); err != nil {
+				return fmt.Errorf("failed to render label: %w", err)
+			}
+		}
+
+		if err := renderFileContentAsMarkdown("output.json", string(marshaled), w, io); err != nil {
+			return fmt.Errorf("failed to render JSON: %w", err)
+		}
+	}
+	return nil
+}
+
+// renderRawMarkdown renders the given raw markdown string to the given writer.
+// Use for complete markdown content from tool calls that need no conversion.
+func renderRawMarkdown(md string, w io.Writer, io *iostreams.IOStreams) error {
+	// Glamour doesn't add leading newlines when content is a complete
+	// markdown document. So, we must add the leading newline.
+	formatFunc := func(s string) string {
+		return fmt.Sprintf("\n%s\n\n", s)
+	}
+
+	return renderMarkdownWithFormat(md, w, io, formatFunc)
+}
+
+// renderMarkdownWithFormat renders the given markdown string to the given writer.
+// If a formatFunc is provided, the md string is ran through it before
+// rendering. This can be used to add newlines before and after the content.
+func renderMarkdownWithFormat(md string, w io.Writer, io *iostreams.IOStreams, formatFunc func(string) string) error {
+	rendered, err := markdown.Render(md,
+		markdown.WithTheme(io.TerminalTheme()),
+		markdown.WithWrap(io.TerminalWidth()),
+	)
+
+	if err != nil {
+		return fmt.Errorf("failed to render markdown: %w", err)
+	}
+
+	rendered = strings.TrimSpace(rendered)
+	if formatFunc != nil {
+		rendered = formatFunc(rendered)
+	}
+
+	fmt.Fprint(w, rendered)
+
+	return nil
+}
+
+// stripDiffFormat implements a primitive conversion from a diff string to a
+// plain text representation by removing diff-specific formatting.
+func stripDiffFormat(diff string) string {
+	lines := strings.Split(diff, "\n")
+
+	// Find where the hunk header ends.
+	hunkEndIndex := -1
+	for i, line := range lines {
+		if strings.HasPrefix(line, "@@") {
+			hunkEndIndex = i
+			break
+		}
+	}
+
+	// If we found the hunk header end, we strip everything before it.
+	if hunkEndIndex != -1 {
+		lines = lines[hunkEndIndex+1:]
+	} else {
+		// This isn't a diff, so we defensively just return the original string.
+		return diff
+	}
+
+	// Now we strip the leading + and - from lines, if they exist.
+	// Note: most of the time, but not all the time, we get a diff without
+	// these prefixes.
+	var stripped []string
+	for _, line := range lines {
+		if strings.HasPrefix(line, "+") || strings.HasPrefix(line, "-") {
+			stripped = append(stripped, line[1:])
+		} else {
+			stripped = append(stripped, line)
+		}
+	}
+	return strings.Join(stripped, "\n")
+}
+
+// renderFileContentAsMarkdown renders the given content as markdown
+// based on the file extension of the path.
+func renderFileContentAsMarkdown(path, content string, w io.Writer, io *iostreams.IOStreams) error {
+	parts := strings.Split(path, ".")
+	lang := parts[len(parts)-1]
```

Comment:
Since `path` can include dots other than the actual file extension, I think we should use the `filepath` package to help us extract the right extension.


```suggestion
	// We need to apply filepath.ToSlash to make sure the behaviour is the same
	// regardless of the platform.
	lang := filepath.Ext(filepath.ToSlash(path))
```